### PR TITLE
feat: add surveyreport questions

### PR DESCRIPTION
**Pull request by Alec4r**
## DESCRIPTION
This PR creates a new hook Action that allows tutor plugins to interact with the configuration at the time of the interactive questionnaire that happens during `tutor local launch`.

## REQUIRED PLUGINS
None. The creation of the Action does not require anything from new or existing plugins. Current plugins that do not use this action will see absolutely no changes to the functionality of `tutor local launch`.

## MORE INFO
This Action was first needed when trying to create a plugin to send automatic and anonymized reports of usage.
For more context on that see:
- https://github.com/eduNEXT/tutor-contrib-surveyreport
- https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3872948238/Problem+Very+few+are+going+to+generate+and+send+the+Survey+Report.

## EXAMPLE:

A plugin using this Action to add a new question can be seen at:
https://github.com/eduNEXT/tutor-contrib-surveyreport/pull/1

### CODE REVIEW

File: tutor/interactive.py [hunk 1]
```diff
@@ -145,6 +145,40 @@ def ask_questions(config: Config, run_for_prod: Optional[bool] = None) -> None:
             config,
             defaults,
         )
+    if plugins.is_loaded("surveyreport"):
```
Comment (felipemontoya) [inline tutor/interactive.py:148]: This PR in the current form is adding to the tutor core a custom way of asking questions that only apply to the survey report plugin. I don't think that would be the way of doing this feature.
Tutor is build on plugability, we need to create a new filter right here that we can later extend from the survey report.

File: tutor/interactive.py [hunk 2]
```diff
@@ -145,6 +145,40 @@ def ask_questions(config: Config, run_for_prod: Optional[bool] = None) -> None:
...
+        ask_survey_report_questions(config=config, defaults=defaults)
+
+
+def ask_survey_report_questions(config: Config, defaults: Config) -> None:
```
Comment (felipemontoya) [inline tutor/interactive.py:152]: This is the function that should go in the survey plugin

File: tutor/templates/config/base.yml [hunk 3]
```diff
@@ -8,6 +8,8 @@ OPENEDX_SECRET_KEY: "{{ 24|random_string }}"
 PLUGINS:
   # The MFE plugin is required
   - mfe
+  # SurveyReport plugin is required
+  - surveyreport
```
Comment (felipemontoya) [inline tutor/templates/config/base.yml:12]: We can not allow the survey report to become a requirement for every tutor install. We have a proposal to make it part of tutor full, but first we need to finish up the plugin and make it completely optiona. Only then we can create a separate request to include it in https://github.com/overhangio/tutor/blob/master/requirements/plugins.txt

**Review by felipemontoya — CHANGES_REQUESTED** [review]
We need to refactor this as a new filter that other plugins can also build on top of

**Comment by felipemontoya** [comment]
@regisb we sat with @Alec4r in a call giving this PR a big refactor and we'd like you input about the ergonomics of the new Filter. This is probably the first filter that is being contributed so we could not find an example to draw from, so please forgive our lack of understanding.

We are looking at adding the filter at the end of the `ask_questions` [function](https://github.com/overhangio/tutor/blob/3b2373f6f14de595200a41966cd27c27369368b5/tutor/interactive.py#L147).

#### Option 1
would be to add something that lets passes the context to the filter and we expect the plugins to run the code that asks the questions. Something like:

at tutor core:
```python
utils = [ask, ask_bool, ask_choice]
config = ASK_INTERACTIVE_QUESTIONS.apply(config, defaults, utils)
```
at the plugin:
```python
@hooks.Filter.ASK_INTERACTIVE_QUESTIONS.add()
def cool_question_function(config, defaults, utils):

    utils.ask("do you want to automatically send a survey report to Axim every six months? pretty please :)", "AUTO_SEND_SURVEY_REPORT", config, defaults)

    return config
```

#### Option 2

would be to a control loop at the core that asks the question that are mostly string configured at the plugin.
Something like:

at tutor core:
```python
 for item in hooks.Filter.ASK_INTERACTIVE_QUESTIONS.get_items():
    ask(item[2], item[3], config, defaults)
```
at the plugin:
```python
hooks.Filters.ASK_INTERACTIVE_QUESTIONS.add_items(
    [
        (
            "surveyreport-namespace",
            "do you want to automatically send a survey report to Axim every six months? pretty please :)",
             "AUTO_SEND_SURVEY_REPORT",
        ),
        (
            "surveyreport-namespace",
            "Are you sure?",
             "AUTO_SEND_SURVEY_REPORT_CONFIRMATION",
        ),
    ]
)
```

Any and all guidance is very appreciated.

**Comment by regisb** [comment]
I think I may have put you on the wrong lead. Instead of a filter, maybe create a `INTERACTIVE_CONFIGURATION` action that would take as argument the current configuration?

```
@hooks.Action.INTERACTIVE_CONFIGURATION.add()
def _enable_surveyreport(config):
    utils.ask("...")
    ...

def ask_questions(config, run_for_prod):
    hooks.Action.INTERACTIVE_CONFIGURATION.do(config)
```

**Comment by Alec4r** [comment]
@regisb @felipemontoya I have added the actions to tutor core and also I openned a PR in surveyreport plugin https://github.com/eduNEXT/tutor-contrib-surveyreport/pull/1 adding the questions.

**Review by felipemontoya — APPROVED** [review]
Thanks a lot @Alec4r. This is now very much in line with what I was expecting we needed to do in the core of tutor.

The intervention seems lightweight and has an explanation at the definition of the new Action. Would you please go ahead and review this @regisb?

@Alec4r I think we need to update the PR description to make clear that this is no longer surveyreport specific but it is only opening up a space for plugins to ask question during the interactive questionnaire.

File: tutor/hooks/catalog.py [hunk 4]
```diff
@@ -123,6 +123,12 @@ def run_this_on_start(root, config, name):
     #: :parameter str root: absolute path to the project root.
     PROJECT_ROOT_READY: Action[str] = Action()
 
+    #: Triggered after all interactive questions have been asked.
+    #: You should use this action if you want to add new questions.
+    #:
+    #: :parameter dict config: project configuration.
+    INTERACTIVE_CONFIGURATION: Action[[Config]] = Action()
```
Comment (regisb) [inline tutor/hooks/catalog.py:130]: Can you please move this action such that it's listed in alphabetical order? Also, we should rename it to `CONFIG_INTERACTIVE`.

**Review by regisb — CHANGES_REQUESTED** [review]
Who doesn't love simplicity? Thanks @Alec4r! Can you please just add a changelog entry?

**Comment by regisb** [comment]
Sorry about the delay guys, I was busy with the Quince release.

File: tutor/hooks/catalog.py (hunk 4 above)
Comment (Alec4r) [reply tutor/hooks/catalog.py]: @regisb Done

**Comment by Alec4r** [comment]
> Who doesn't love simplicity? Thanks @Alec4r! Can you please just add a changelog entry?

I think is done, let me know if you have another comment.

**Review by regisb — APPROVED** [review]

**Pull request merged**